Sources/Bookmarks/BookmarkEditorViewModel.swift

```diff
@@ -0,0 +1,179 @@
+//
+//  BookmarkEditorViewModel.swift
+//
+//  Copyright © 2022 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Combine
+import CoreData
+import Persistence
+import Common
+
+public class BookmarkEditorViewModel: ObservableObject {
+
+    public struct Location {
+
+        public let bookmark: BookmarkEntity
+        public let depth: Int
+    }
+
+    let context: NSManagedObjectContext
+
+    @Published public var bookmark: BookmarkEntity
+    @Published public var locations = [Location]()
+
+    lazy var favoritesFolder: BookmarkEntity! = BookmarkUtils.fetchFavoritesFolder(context)
+
+    private let subject = PassthroughSubject<Void, Never>()
+    public var externalUpdates: AnyPublisher<Void, Never>
+    
+    private let errorEvents: EventMapping<BookmarksModelError>?
+
+    public var canSave: Bool {
+        let titleOK = bookmark.title?.trimmingWhitespace().count ?? 0 > 0
+        let urlOK = bookmark.isFolder ? true : bookmark.urlObject != nil
```

Comment:
This will reject bookmarklets, which in turn will break the Fingerprinting UI tests as they won't be able to save the edited bookmark

Reply:
OK, i've pushed a fix for this (to both iOS and BSK)

Reply:
Alright, this looks good now